### PR TITLE
Chore: Deprecate FieldConfigEditorProps in favor of StandardEditorProps

### PR DESCRIPTION
**Pull request by joshhunt**
This PR deprecates `FieldConfigEditorProps` and aliases it to `StandardEditorProps`, and updates our editors to use `StandardEditorProps` everywhere.

It also stops the recursive type reference (the recursiveness itself wasn't a problem) that leaked too much 'implementation detail' of this system to an external API. I removed the `StandardFieldItem` -> `StandardFieldEditor` -> `StandardFieldItem` references to instead just inline the parts of `StandardFieldItem` that editors use (mainly just `settings`)

There is no change in runtime behaviour from this change, and aliasing `FieldConfigEditorProps` should drastically limit any typescript 'breaking changes' from this deprecation. 

Replaces work initially done in https://github.com/grafana/grafana/pull/78054 

---

In reviewing the current panel plugin options UI builders, we discovered there's a few layers of type loss that makes it difficult to understand that the correct type that panel options and field config editors have. 

For context, there are two types available for panel options/field config:
 - `StandardEditorProps` is the base type
 - `FieldConfigEditorProps` [extends `StandardEditorProps`](https://github.com/grafana/grafana/blob/5f6fea33f293ac57625305afa74027654716f452/packages/grafana-data/src/types/fieldOverrides.ts#L76) mainly to extend the context property to add two additional optional properties that are not used in our codebase

Standard editor components are defined with the `FieldConfigEditorProps` type, despite not using any of the extra properties:

https://github.com/grafana/grafana/blob/5f6fea33f293ac57625305afa74027654716f452/public/app/core/components/OptionsUI/number.tsx#L7-L9

But then they're put into a type that expects it to be `StandardEditorProps`, so `as any` is used to squeeze it in:

https://github.com/grafana/grafana/blob/5f6fea33f293ac57625305afa74027654716f452/public/app/core/components/OptionsUI/registry.tsx#L54-L59

Then those registry items are squeezed into another type that expects them to be `FieldConfigEditorProps`, with `as any` again destroying any hint of what the expected type should be

https://github.com/grafana/grafana/blob/5f6fea33f293ac57625305afa74027654716f452/packages/grafana-data/src/utils/OptionsUIBuilders.ts#L27-L41

Note these are also used in a type that expects them to be `StandardEditorProps` (which they are), but they're needlessly squeezed through `as any`

https://github.com/grafana/grafana/blob/5f6fea33f293ac57625305afa74027654716f452/packages/grafana-data/src/utils/OptionsUIBuilders.ts#L224-L230

After reviewing our first party use of these editors, we discovered that the additional context properties aren't even passed into field config override editors. There [should/could be `field` and dataFrameIndex` properties](https://github.com/grafana/grafana/blob/5f6fea33f293ac57625305afa74027654716f452/packages/grafana-data/src/types/fieldOverrides.ts#L72-L75), but they're not actually passed in at runtime:

6485_2023-11-17-11-39_firefox

https://github.com/grafana/grafana/blob/53186c14a4ccac9b360656d3f46a1892c2b45c05/public/app/features/dashboard/components/PanelEditor/getFieldOverrideElements.tsx#L70-L74

### CODE REVIEW

**Comment by joshhunt** [comment]
I've done a review of _known_ public external plugins, and I've confirmed that no one is using the additional context properties (that are not there in the first place).

Only three plugins use the `Field*EditorProps` interfaces, and I've reviewed them and confirm they should be compatible with this change:

 - [alexandra-trackmap-panel/colorMapEditor.tsx](https://github.com/alexandrainst/alexandra-trackmap-panel/blob/b89f6fbd99dc6b0e99464d0c75a0f648d5675401/src/colorMapEditor.tsx#L16)
 - [alexandra-trackmap-panel/stringMapEditor.tsx](https://github.com/alexandrainst/alexandra-trackmap-panel/blob/b89f6fbd99dc6b0e99464d0c75a0f648d5675401/src/stringMapEditor.tsx#L16)
 - [grafana-lenav-panel/DataLinksValueEditor.tsx](https://github.com/lework/grafana-lenav-panel/blob/f2f67a9f83fe5242c0584ff7382211ec9525d4ad/src/DataLinksInlineEditor/DataLinksValueEditor.tsx#L6)
 - [Grafana_Status_panel/StatusThresholdOptionsEditor.tsx](https://github.com/Vonage/Grafana_Status_panel/blob/3e344e4042662e74db6c04c8c54b5be1afe83f39/src/components/StatusThresholdOptionsEditor.tsx#L42)

**Comment by mckn** [comment]
> I've done a review of _known_ public external plugins, and I've confirmed that no one is using the additional context properties (that are not there in the first place).
> 
> Only three plugins use the `Field*EditorProps` interfaces, and I've reviewed them and confirm they should be compatible with this change:
> 
> * [alexandra-trackmap-panel/colorMapEditor.tsx](https://github.com/alexandrainst/alexandra-trackmap-panel/blob/b89f6fbd99dc6b0e99464d0c75a0f648d5675401/src/colorMapEditor.tsx#L16)
> * [alexandra-trackmap-panel/stringMapEditor.tsx](https://github.com/alexandrainst/alexandra-trackmap-panel/blob/b89f6fbd99dc6b0e99464d0c75a0f648d5675401/src/stringMapEditor.tsx#L16)
> * [grafana-lenav-panel/DataLinksValueEditor.tsx](https://github.com/lework/grafana-lenav-panel/blob/f2f67a9f83fe5242c0584ff7382211ec9525d4ad/src/DataLinksInlineEditor/DataLinksValueEditor.tsx#L6)
> * [Grafana_Status_panel/StatusThresholdOptionsEditor.tsx](https://github.com/Vonage/Grafana_Status_panel/blob/3e344e4042662e74db6c04c8c54b5be1afe83f39/src/components/StatusThresholdOptionsEditor.tsx#L42)

Nice, thanks for the extensive checks on this 🙏🏻

**Comment by academo** [comment]
>Only three plugins use the Field*EditorProps interfaces, and I've reviewed them and confirm they should be compatible with this change:

There's no guarantee that only these 3 plugins are using the props. A better description is: We could only find 3 publicly available plugin's source code using these fields.

**Comment by joshhunt** [comment]
Right - that was what I meant by reviewing "_known_" plugins. The review of our internal plugins, and submitted open source plugins showed me that there's _little_ value (if any) of the properties I'm "hiding" by changing the types. I used this as a signal that this indicate that this change is probably inline with most developer's expectations and needs.

Note that I'm not actually changing any runtime code (so no plugins should break as Grafana is updated), and I'm aliasing the deprecated type to the new one to maximise build-time compat.

**Comment by joshhunt** [comment]
@dprokop flagged to me that the `FieldOverrideContext` context - which we no longer show is exposed to editors through the `item` prop - is used in some 'processors'

https://github.com/grafana/grafana/blob/91a5c3803c3824108ad768cea227cf8527eed684/packages/grafana-data/src/field/overrides/processors.ts#L35-L43

This Is Fine Actually. We're only changing the type of what's passed to _editors_, not what's passed to the `process` functions. The [code path that calls this](https://github.com/grafana/grafana/blob/91a5c3803c3824108ad768cea227cf8527eed684/packages/grafana-data/src/field/fieldOverrides.ts)  remains reasonably well typed, and any change in the types would show errors here.

File: packages/grafana-data/src/field/standardFieldConfigEditorRegistry.ts [hunk 1]
```diff
@@ -19,10 +19,14 @@ export interface StandardEditorContext<TOptions, TState = any> {
 export interface StandardEditorProps<TValue = any, TSettings = any, TOptions = any, TState = any> {
   value: TValue;
   onChange: (value?: TValue) => void;
-  item: StandardEditorsRegistryItem<TValue, TSettings>;
   context: StandardEditorContext<TOptions, TState>;
   id?: string;
+
+  item: RegistryItem & {
+    settings?: TSettings;
+  };
```
Comment (dprokop) [inline packages/grafana-data/src/field/standardFieldConfigEditorRegistry.ts:27]: 👍

**Review by dprokop — APPROVED** [review]
@joshhunt - great effort getting these types straight. 

I would suggest getting a second pair of eyes from the @grafana/dataviz-squad and this should be good to merge.

**Review by ashharrison90 — APPROVED** [review]
🔥🔥🔥🔥🔥

**Pull request merged**